src/mame/yamaha/ymdx9.cpp

```diff
@@ -0,0 +1,449 @@
+	 */
+	uint8_t p1_r(offs_t offset);
+
+	void unmapped_w(offs_t address, uint8_t data);
+
+	uint8_t unmapped_r(offs_t offset);
+};
+
+
+/**
+* yamaha_dx9_state::machine_start
+*/
+void yamaha_dx9_state::machine_start()
+{
+	m_leds.resolve();
+	m_rx_data = ASSERT_LINE;
+}
+
+
+/**
+ * yamaha_dx9_state::lcd_pixel_update
+ */
+HD44780_PIXEL_UPDATE(yamaha_dx9_state::lcd_pixel_update)
+{
+	if (x < 5 && y < 8 && line < 2 && pos < 16)
+			bitmap.pix(line * 10 + y + 1 + ((y == 7) ? 1 : 0), pos * 6 + x + 1) = state ? 1 : 2;
+}
+
+
+/**
+ * yamaha_dx9_state::palette_init
+ */
+void yamaha_dx9_state::palette_init(palette_device &palette)
+{
+	palette.set_pen_color(0, rgb_t(0x87, 0xAD, 0x34)); // background
+	palette.set_pen_color(1, rgb_t(0x0, 0x0, 0x0)); // lcd pixel on
+	palette.set_pen_color(2, rgb_t(0x7D, 0x9F, 0x32)); // lcd pixel off
+}
+
+
+/**
+ * yamaha_dx9_state::mem_map
+ */
+void yamaha_dx9_state::mem_map(address_map &map)
+{
+	// Internal CPU registers.
+	map(0x0000, 0x001f).m(m_maincpu, FUNC(hd6303r_cpu_device::m6801_io));
+
+	map(0x20, 0x20).r(FUNC(yamaha_dx9_state::key_switch_scan_driver_r));
+
+	map(0x22, 0x22).r("adc", FUNC(m58990_device::data_r));
+	map(0x24, 0x24).w("adc", FUNC(m58990_device::address_data_start_w));
+
+	// YM21280 OPS.
+	map(0x26, 0x27).w(FUNC(yamaha_dx9_state::ops_w));
+	// HD44780 LCD Controller.
+	map(0x0028, 0x0029).rw("lcdc", FUNC(hd44780_device::read), FUNC(hd44780_device::write));
+	// LED.
+	map(0x2B, 0x2C).w(FUNC(yamaha_dx9_state::led_w));
+
+	// Internal RAM.
+	map(0x0040, 0x00ff).ram();
+
+	// External RAM.
+	// 2 * 2kb RAM1 IC19 M5M118P.
+	map(0x0800, 0x0fff).ram().share("ram1");
+	map(0x1000, 0x1800).ram().share("ram2");
+
+	// YM21290 EGS
+	map(0x1800, 0x18F3).w(FUNC(yamaha_dx9_state::egs_w));
+
+	// ROM.
+	map(0xc000, 0xffff).rom().region("program", 0);
+}
+
+
+/**
+ * yamaha_dx9_state::dx9
+ */
+void yamaha_dx9_state::dx9(machine_config &config)
+{
+	// Initialise the HD63B03RP CPU.
+	// This oscillator frequency comes from the service manual.
+	HD6303R(config, m_maincpu, 9.4265_MHz_XTAL / 2);
+	m_maincpu->set_addrmap(AS_PROGRAM, &yamaha_dx9_state::mem_map);
+
+	// Unlike the DX7 only IO port 1 is used.
+	// The direction flags of other ports are set, however they are never read, or written.
+	m_maincpu->in_p1_cb().set(FUNC(yamaha_dx9_state::p1_r));
+	m_maincpu->in_p2_cb().set(FUNC(yamaha_dx9_state::unmapped_r));
+	m_maincpu->in_p3_cb().set(FUNC(yamaha_dx9_state::unmapped_r));
+	m_maincpu->in_p4_cb().set(FUNC(yamaha_dx9_state::unmapped_r));
+
+	m_maincpu->out_p1_cb().set(FUNC(yamaha_dx9_state::p1_w));
+	m_maincpu->out_p2_cb().set(FUNC(yamaha_dx9_state::unmapped_w));
+	m_maincpu->out_p3_cb().set(FUNC(yamaha_dx9_state::unmapped_w));
+	m_maincpu->out_p4_cb().set(FUNC(yamaha_dx9_state::unmapped_w));
+
+	NVRAM(config, "ram1", nvram_device::DEFAULT_ALL_0);
+	NVRAM(config, "ram2", nvram_device::DEFAULT_ALL_0);
+
+	// Configure the ADC. The clock speed here is a guess.
+	M58990(config, m_adc, 8_MHz_XTAL / 16);
```

Comment:
If the clock speed is only guessed don't use a specific XTAL value here.

Reply:
No problem. Do you have a suggestion of what I should put here instead?

Reply:
You can just use a literal integer like `500'000` – we use literal numbers when we don’t know what the clock source is.